File: clients/php/BufferedStream.php
<?php

require_once 'Stream.php';

define('BUFFER_SIZE', 102_400);

class BufferedInputStream extends InputStream
{
    private InputStream $inner;
    private string $buffer;
    private int $bufferPos;
    function __construct(InputStream $inner)
    {
        $this->inner = $inner;
        $this->buffer = '';
        $this->bufferPos = 0;
    }
    function readAtMost(int $byteCount): string
    {
        if ($this->bufferPos == strlen($this->buffer)) {
            $this->buffer = $this->inner->readAtMost(BUFFER_SIZE);
            $this->bufferPos = 0;
        }
        $chunk = substr($this->buffer, $this->bufferPos, $byteCount);
        $this->bufferPos += strlen($chunk);
        return $chunk;
    }
}

class BufferedOutputStream extends OutputStream
{
    private OutputStream $inner;
    private string $buffer;
    function __construct(OutputStream $inner)
    {
        $this->inner = $inner;
        $this->buffer = '';
    }
    function write(string $bytes): void
    {
        $this->buffer .= $bytes;
    }
    function flush(): void
    {
        $this->inner->write($this->buffer);
        $this->inner->flush();
        $this->buffer = '';
    }
}

File: clients/swift/Sources/AiCup22/BufferedStream.swift
let BUFFER_SIZE = 102_400

class BufferedInputStream : InputStream {
    init(_ inner: InputStream) {
        self.inner = inner
        buffer = []
        bufferPos = 0
    }

    func readBytesAtMost(_ byteCount: Int) -> [Byte] {
        if bufferPos == buffer.count {
            buffer = inner.readBytesAtMost(BUFFER_SIZE);
            bufferPos = 0;
        }
        let chunk = Array(buffer[bufferPos..<min(bufferPos + byteCount, buffer.count)]);
        bufferPos += chunk.count;
        return chunk;
    }

    private let inner: InputStream
    private var buffer: [Byte]
    private var bufferPos: Int
}

class BufferedOutputStream : OutputStream {
    init(_ inner: OutputStream) {
        self.inner = inner
        buffer = []
    }

    func writeBytes(_ data: [Byte]) {
        self.buffer.append(contentsOf: data)
    }

    func flush() {
        self.inner.writeBytes(self.buffer)
        self.inner.flush()
        self.buffer = []
    }
    
    private let inner: OutputStream
    private var buffer: [Byte]
}